### PR TITLE
[WIP] Weina

### PR DESCRIPTION
**Pull request by jhprinz**
Add the alanine example from

> [1] W.-N. Du, K. A. Marino, and P. G. Bolhuis, “Multiple state transition interface sampling of alanine dipeptide in explicit solvent,” J. Chem. Phys., vol. 135, no. 14, p. 145102, 2011.

The current setting is
- Alanine Dipeptide (22 atoms)
- in Waterbox ~ 2.6 nm with (543 water molecules)
- total n_atoms 1651
- openmm setting with Langevin integrator (I think this should be changed)
- amber96 FF (like in the original paper) using a newer FF shifts the centers of the states so much that these are not metastable anymore.
- water FF tip3P

```
forcefield = app.ForceField('amber96.xml', 'tip3p.xml')

omm_topology = paths.tools.to_openmm_topology(template)

system = forcefield.createSystem(omm_topology, nonbondedMethod=app.PME, 
    nonbondedCutoff=1.0*unit.nanometers, constraints=app.HBonds, rigidWater=True, 
    ewaldErrorTolerance=0.0005)
integrator = mm.LangevinIntegrator(300*unit.kelvin, 1.0/unit.picoseconds, 
    2.0*unit.femtoseconds)
integrator.setConstraintTolerance(0.00001)
```

The example has three notebooks

1. `weina_bootstrap.ipynb`

    Contains finding initial configurations in each of the 6 states and then run a full bootstrap in each state
    Finally save all necessary data in one file and create a combines sampleset.

2. `weina_production.ipynb`
    Loading from bootstrap, construct minus ensembles (needs improvement) and do production moves

3. `weina_analysis.ipynb`
    Analysis in analogy to the `mstis_analysis.ipynb` notebook

### CODE REVIEW

**Comment by jhprinz** [comment]
So far this works and I will run a long simulation ~ 15000 steps tonight. It took me some time to figure out that I had to adjust the binning but it is clear now.

To get the initial configuration I used a VERY high temperature so far and in some cases using these leads to `NaN` in the simulation. So we may have to reduce this, but otherwise it is working.

We should have a close look if all moves are doing what they should. I had problems to construct initial minus samples. The implemented function returned a trajectory that I needed to use `.split` on to get a valid minus sample. After that everything worked. Still not sure if this causes a problem. I realized that some RepEx moves have zero acceptance
```

ReplicaExchange ran 1.24444444444% of the cycles with acceptance 15/28 (0.535714285714)
ReplicaExchange ran 1.42222222222% of the cycles with acceptance 13/32 (0.40625)
ReplicaExchange ran 1.42222222222% of the cycles with acceptance 0/32 (0.0)
ReplicaExchange ran 1.2% of the cycles with acceptance 18/27 (0.666666666667)
ReplicaExchange ran 1.24444444444% of the cycles with acceptance 26/28 (0.928571428571)
ReplicaExchange ran 1.06666666667% of the cycles with acceptance 7/24 (0.291666666667)
ReplicaExchange ran 1.28888888889% of the cycles with acceptance 4/29 (0.137931034483)
ReplicaExchange ran 1.95555555556% of the cycles with acceptance 10/44 (0.227272727273)
ReplicaExchange ran 1.51111111111% of the cycles with acceptance 12/34 (0.352941176471)
ReplicaExchange ran 1.55555555556% of the cycles with acceptance 14/35 (0.4)
ReplicaExchange ran 1.55555555556% of the cycles with acceptance 22/35 (0.628571428571)
ReplicaExchange ran 1.55555555556% of the cycles with acceptance 33/35 (0.942857142857)
ReplicaExchange ran 1.33333333333% of the cycles with acceptance 0/30 (0.0)
ReplicaExchange ran 1.33333333333% of the cycles with acceptance 20/30 (0.666666666667)
ReplicaExchange ran 1.15555555556% of the cycles with acceptance 0/26 (0.0)
ReplicaExchange ran 1.06666666667% of the cycles with acceptance 10/24 (0.416666666667)
```

and one path reversal always failed
```
PathReversal ran 1.82222222222% of the cycles with acceptance 28/41 (0.682926829268)
PathReversal ran 1.51111111111% of the cycles with acceptance 23/34 (0.676470588235)
PathReversal ran 1.68888888889% of the cycles with acceptance 31/38 (0.815789473684)
PathReversal ran 1.06666666667% of the cycles with acceptance 6/24 (0.25)
PathReversal ran 1.33333333333% of the cycles with acceptance 27/30 (0.9)
PathReversal ran 1.06666666667% of the cycles with acceptance 15/24 (0.625)
PathReversal ran 1.24444444444% of the cycles with acceptance 25/28 (0.892857142857)
PathReversal ran 1.46666666667% of the cycles with acceptance 26/33 (0.787878787879)
PathReversal ran 1.46666666667% of the cycles with acceptance 26/33 (0.787878787879)
PathReversal ran 1.73333333333% of the cycles with acceptance 0/39 (0.0)
PathReversal ran 2.04444444444% of the cycles with acceptance 43/46 (0.934782608696)
PathReversal ran 1.6% of the cycles with acceptance 17/36 (0.472222222222)
PathReversal ran 1.91111111111% of the cycles with acceptance 41/43 (0.953488372093)
PathReversal ran 1.64444444444% of the cycles with acceptance 23/37 (0.621621621622)
PathReversal ran 1.2% of the cycles with acceptance 27/27 (1.0)
PathReversal ran 1.55555555556% of the cycles with acceptance 35/35 (1.0)
PathReversal ran 1.37777777778% of the cycles with acceptance 14/31 (0.451612903226)
```

**Comment by dwhswenson** [comment]
This looks like a great start for this. Let's change the name to "alanine_dipeptide" or "ala" or something; both because the current name will be meaningless to a user (whereas these notebooks will be really useful as guides for how to set things up) and because @WeinaUvA is (and did) much more than alanine dipeptide. (I don't think she'll get the notification -- I noticed she's still on the OPS team, though).

> I realized that some RepEx moves have zero acceptance...
> ...and one path reversal always failed

Replay the simulation using the live visualization tools. My guess: you'll see that one ensemble (the one which always rejects path reversal) always has an A->B path in it. Over 10s of shooting moves per ensemble, that's not surprising. Same with the repex pairs that don't allow swaps. Path sampling takes a while to equilibrate. Worry about it if your results tomorrow show no success (assuming you have ~100s or shooting moves per ensemble).

> I had problems to construct initial minus samples.

This is more of a concern. However, I'm a little confused, just glancing at this: why is there even a possibility of populating `I`? Usually, we fix the innermost interface to be at exactly the state boundary. If the state labels are what they claim to be in cell 19 of `weina_production`, that's the expected behavior -- a minus trajectory should start and end with `A`, have one other `A` in between, with at least one `X` between each `A`.

> So far this works and I will run a long simulation ~ 15000 steps tonight.

Aside: it can be easier to think about these things in terms of number of expected shooting moves per ensemble -- the number of total steps depends on the number of ensembles, and so is a bit harder to imagine. Check out the `MoveScheme.n_steps_for_trials` and `MoveScheme.n_trials_for_steps` functions. 

Second aside: I use different colors for the paths in different ensembles because it really does help see what's going on -- your plot with all the bootstrapped paths is pretty, but I can't tell one path from another, since they're all black.

**Comment by dwhswenson** [comment]
Another renaming thought: rename the `_production.ipynb` and `_analysis.ipynb` to include `mstis`. We'll use the same TIS network for SRTIS, too, just using an SRTIS move scheme instead (very close to finished).

**Comment by jhprinz** [comment]
Thanks for the input. Yes the unaccepted reversal was indeed A->B path problem and it occurred only in one state. Currently running the 15000 steps and seems to work after my power saving stopped after an hour. 
I wrapped the production in a try except because at the high energy states the `2fs` steps seemed to be too large and could cause `NaN`s. in that cause the current step is aborted and restarted. The number 15000 means 15000 successful completed steps.

To summarize the performance a little (rough estimates).
- 35 ensembles total
- ~ 5000 MC steps / hour
- File size : 11,1 GB
- Time to open the file including caching to memory : 90s
- File stats

    PathMovers: 187
    Engines: 1
    Samples: 19575
    Ensembles: 411
    SampleSets: 15000
    Snapshots: 545072
    Configurations: 272536
    Trajectories: 11847
    Networks: 1

- time to complete rate matrix 6x6: 142s

**Comment by bolhuis** [comment]
Can this be merged? I would like to have a look, if possible...

**Comment by dwhswenson** [comment]
This is not ready to merge, especially since the last version failed tests (probably due to a problem with insufficient data in one of the tests, not a real fundamental failure, but still, don't merge things that fail tests.) But there are ways to access it without it being merged to the main repository.

Follow these instructions to add @jhprinz's fork as a new remote source: https://help.github.com/articles/adding-a-remote/

I would recommend naming it something like "jhprinz" or "jhp". Example:
```
git remote add jhp https://github.com/jhprinz/openpathsampling.git
```

Then you can check out any branch from his fork just as you can check out any branch of the main. To get this branch, do:

```
git fetch jhp
git checkout jhp/weina
```

Then you will be in the most recent version of this branch. To get back to your own master branch, just type `git checkout master` (although you'll probably need to discard changes)

**Comment by jhprinz** [comment]
I will try to fix most of it to get this passing and we can then improve on the nb content. This is not too difficult. I restarted travis to see if it was just bad luck of insufficient data.

It changes BootStrapping a little (hope this is an improvement) and a little change in the live vis to make this work with my phi/psi background picture. 

The rest is just adding the weina notebooks (still need to rename it). So it should not affect any existing code and tests.

**Comment by bolhuis** [comment]
I have followed your instructions and made branch for testing the AD example. I run into the following error at step 4 of the bootstrap:

platform = mm.Platform.getPlatformByName('OpenCL')

which returns:

---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-4-22c4f17f756f> in <module>()
     10 integrator.setConstraintTolerance(0.00001)
     11 
---> 12 platform = mm.Platform.getPlatformByName('OpenCL')
     13 properties = {'OpenCLPrecision': 'mixed'}

/Users/bolhuis/packages/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.pyc in getPlatformByName(*args)
  13384 
  13385 
> 13386         return _openmm.Platform_getPlatformByName(*args)
  13387 
  13388     if _newclass:getPlatformByName = staticmethod(getPlatformByName)

Exception: There is no registered Platform called "OpenCL"

Do I have to install something? I am using a Macbook Pro with, Intel Iris gpu.

**Comment by dwhswenson** [comment]
@bolhuis : For some reason, the engine in this is hard-wired to OpenCL. I seem to remember that there was a period that OpenMM's OpenCL platform didn't work on (some?) Macs, although I thought that had been fixed (@jchodera or @jhprinz will know more than I do). First thing to try would probably be to make sure you're using the most recent version of OpenMM.

Worst case, you can run it on the CPU platform, but that will be painfully slow. For that, just change the `'OpenCL'` to `'CPU'` (I think that's right -- is platform case-sensitive?). Anyway, the others can help you with this more than I can.

@jhprinz : Wasn't there a "fastest" option for OpenMM platform? Or some way to ***not*** hard-wire it? Hard-wiring the platform in an example we expect users to play with seems like a really bad idea, since not all platforms will be available to all users.

**Comment by jchodera** [comment]
If you just omit the `platform` argument from
```python
context = openmm.Context(system, integrator, platform)
```
and instead use 
```python
context = openmm.Context(system, integrator)
```
it will automatically pick the fastest available platform.

**Comment by bolhuis** [comment]
I got it to work in 'CPU' mode, but it is pretty slow. It looks awesome though.
The macbook pro i have has a IrisIntel and if I test for my gpu using clGetDeviceInfo() I get

Device Intel(R) Core(TM) i7-4750HQ CPU @ 2.00GHz supports OpenCL 1.2 
Device Iris Pro supports OpenCL 1.2 

Surely this should work with OpenCL no? Yet, OpenMM has registered only 2 platforms, 0 is the Reference, 1 CPU. Clearly there is something wrong. 
The version number openmm  is 6.3. That should be up to date.

Any ideas?

**Comment by jchodera** [comment]
It *should* support OpenCL on the latest `osx` version.  What OS version do you have?

**Comment by jchodera** [comment]
Try this in Python:
```python
from simtk import openmm
openmm.Platform.getPluginLoadFailures()
```

**Comment by bolhuis** [comment]
When I try

from simtk import openmm
openmm.Platform.getPluginLoadFailures()

I get 

('Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCUDA.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCUDA.dylib, 9): Library not loaded: @rpath/CUDA.framework/Versions/A/CUDA\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCUDA.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDCUDA.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDCUDA.dylib, 9): Library not loaded: @rpath/CUDA.framework/Versions/A/CUDA\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDCUDA.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeCUDA.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeCUDA.dylib, 9): Library not loaded: @rpath/CUDA.framework/Versions/A/CUDA\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeCUDA.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDOpenCL.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDOpenCL.dylib, 9): Library not loaded: @rpath/libOpenMMOpenCL.dylib\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMRPMDOpenCL.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMAmoebaCUDA.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMAmoebaCUDA.dylib, 9): Library not loaded: @rpath/CUDA.framework/Versions/A/CUDA\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMAmoebaCUDA.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeOpenCL.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeOpenCL.dylib, 9): Library not loaded: @rpath/libOpenMMOpenCL.dylib\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMDrudeOpenCL.dylib\n  Reason: image not found',
 'Error loading library /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCudaCompiler.dylib: dlopen(/Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCudaCompiler.dylib, 9): Library not loaded: @rpath/CUDA.framework/Versions/A/CUDA\n  Referenced from: /Users/bolhuis/packages/miniconda/pkgs/openmm-6.3-py27_1/lib/plugins/libOpenMMCudaCompiler.dylib\n  Reason: image not found')


It seems that it cannot load the libraries. The dylib itself is there, I cheked this, so it seems that the @rpath/libOpenMMOpenCL.dylib is causing the problem. How to check @rpath? I did set the OPENMM_PLUGIN_DIR to the right lib directory, but this seems to have no effect.

**Comment by bolhuis** [comment]
When running the Weina bootstrap OpenMM crashes many times at the first high T run, with the message :"particle coordinate nan". At first I interpreted this as that the temperature is too high, and tried to lower it. However, even at 300 K it does crash.

Here is the message.

Exception                                 Traceback (most recent call last)
<ipython-input-39-79f854d66f52> in <module>()
      7         try:
      8             count += 1
----> 9             tt = engine_high.generate_forward(tt[-1].reversed, core2core)
     10             storage.save(tt)
     11 

/Users/bolhuis/git/openpathsampling/openpathsampling/dynamics_engine.pyc in generate_forward(self, snapshot, ensemble)
    230         """
    231 
--> 232         return self.generate(snapshot, ensemble.can_append, direction=+1)
    233 
    234     def generate_backward(self, snapshot, ensemble):

/Users/bolhuis/git/openpathsampling/openpathsampling/dynamics_engine.pyc in generate(self, snapshot, running, direction)
    305 
    306             # Do integrator x steps
--> 307             snapshot = self.generate_next_frame()
    308             frame += 1
    309             if frame % log_freq == 0:

/Users/bolhuis/git/openpathsampling/openpathsampling/openmm_engine.pyc in generate_next_frame(self)
    226 
    227     def generate_next_frame(self):
--> 228         self.simulation.step(self.nsteps_per_frame)
    229         self._current_snapshot = None
    230         return self.current_snapshot

/Users/bolhuis/packages/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.pyc in step(self, steps)
     93     def step(self, steps):
     94         """Advance the simulation by integrating a specified number of time steps."""
---> 95         self._simulate(endStep=self.currentStep+steps)
     96 
     97     def runForClockTime(self, time, checkpointFile=None, stateFile=None, checkpointInterval=None):

/Users/bolhuis/packages/miniconda/lib/python2.7/site-packages/simtk/openmm/app/simulation.pyc in _simulate(self, endStep, endTime)
    150                 if endTime is not None and datetime.now() >= endTime:
    151                     return
--> 152             self.integrator.step(stepsToGo)
    153             self.currentStep += nextSteps
    154             if anyReport:

/Users/bolhuis/packages/miniconda/lib/python2.7/site-packages/simtk/openmm/openmm.pyc in step(self, *args)
  14263 
  14264 
> 14265         return _openmm.LangevinIntegrator_step(self, *args)
  14266 
  14267 

Exception: Particle coordinate is nan

**Comment by jchodera** [comment]
@bolhuis: How are you installing OpenMM?  Are you using the `conda` installer with `anaconda` or [`miniconda`](http://conda.pydata.org/miniconda.html)?
```
conda config --add channels omnia
conda install --yes openmm
```

**Comment by bolhuis** [comment]
With miniconda. Should I reinstall with anaconda? 

P

**Comment by jhprinz** [comment]
Miniconda and Anaconda are equivalent in this respect. So that should not matter. Did you install openm before not using conda? There might be some _leftovers_ that interfere. This happened to me before.

About the `particle coordinate nan` problem. This just means the simulation exploded. This happened to me too, but only at `1000K` not at room temperature. The notebook is designed to just retry after this happens at `1000K`. This is just until initial configurations in all states have been found. From that moment on all simulations are run at `300K` and so is the bootstrapping in each state.

Did you make sure that this is the case also at `300K`? I presume that reducing the simulation step to `1fs` should fix this. otherwise I am not sure.

**Comment by bolhuis** [comment]
I did install openmm before conda indeed. How do i fix that? Remove all openmm and miniconda and reinstall?

About the crash, i think i first should reinstall. To be continued...

**Comment by jhprinz** [comment]
This is ready for review and merge when tests pass. There are very little adjustments to some parts of openmm engine for convenience and a few additional fail saves for FullBootstrap. Other than that it just add the weina notebooks for the publication.

**Comment by dwhswenson** [comment]
Haha -- and I was just about to check out a copy of this branch as a starting point for the TPS version, which I would then start a PR to master on. I thought it would be funny to take your fork as a starting point, pass it through my fork, and then point the PR to the choderalab fork.

Alas, looks like I'll have to do that on some future idea, since, at first glance, this looks pretty much ready to merge. (I'll take a more careful look at it later.)

One thing: I suggest changing the filenames on these. `weina` doesn't mean anything to the general public. Call it `alanine_dipeptide` or `AD`.

**Comment by jhprinz** [comment]
Very good. Then I was on time. :) Well, yes, changing the name makes sense. I will do that and update the `README.md`.

**Comment by jhprinz** [comment]
Done! Ready to merge when it passes (which it should since the changes did not affect testing !)

**Comment by dwhswenson** [comment]
This works -- merging!

There's some work to be done to clean up for readability, and I don't think you really want to exactly copy the analysis from the toy models (among other things, that's a 3-state model, this is a 6-state model, so reducing to only 3 states is really weird). But the hard part is done; the rest is just about polishing it. I'll add some of that polish soon.

> Very good. Then I was on time. :)

Actually, you were too fast ;-) I was looking forward to using some of the cool tricks that come from git being a distributed VCS, and now I don't need to!
